src/Npgsql/Internal/Converters/RecordObjectArrayConverter.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Internal.Descriptors;
+using Npgsql.PostgresTypes;
+using Npgsql.Properties;
+
+namespace Npgsql.Internal.Converters;
+
+public class RecordObjectArrayConverter : PgStreamingConverter<object[]>
+{
+    readonly PgSerializerOptions _serializerOptions;
+
+    public RecordObjectArrayConverter(PgSerializerOptions serializerOptions)
+        => _serializerOptions = serializerOptions;
+
+    public override object[] Read(PgReader reader)
+        => Read(async: false, reader, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask<object[]> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => Read(async: true, reader, cancellationToken);
+
+    async ValueTask<object[]> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(int)))
+            await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+        var fieldCount = reader.ReadInt32();
+        var result = new object[fieldCount];
+
+        for (var i = 0; i < fieldCount; i++)
+        {
+            if (reader.ShouldBuffer(sizeof(uint) + sizeof(int)))
+                await reader.BufferData(async, sizeof(uint) + sizeof(int), cancellationToken).ConfigureAwait(false);
+
+            var typeOID = reader.ReadUInt32();
+            var fieldLen = reader.ReadInt32();
+
+            // Note that we leave .NET nulls in the object array rather than DBNull.
```

Comment:
Should we?

Reply:
That's the current behavior, not sure there's a good reason to break it... Please in other typed/generic contexts (e.g. array/composite elements) we can't use DBNull anyway, and use null - so this is consistent with that... Let me know if you think we should do something.

Reply:
I think we should, record is effectively the column typed version of a row, and for GetValues we do fill in DBNull

Reply:
I can see it both ways - being consistent with GetValues or with our other compositional types (which admittedly aren't non-generic).

You really think it's worth a break?

Reply:
>You really think it's worth a break?

That I'm not sure of, also not sure whether it'd be worth an appcontext switch... I'll let you decide, if we redid this today I would have written DBNulls though.

Reply:
I'm honestly not entirely sure what I'd do if we redid this today.. DBNull should generally become less important going forward, assuming we fully do generic DbParameter etc. It just doesn't make sense in a typed generic world...

Reply:
IMO all that matters to look at when deciding whether something should be DBNull or not is whether the type of the value is object.

Reply:
It's a possible way to look at it. The one place where I think it definitely makes sense is where there's a distinction to be made between DBNull and null, e.g. the return value of ExecuteScalar. Beyond that I'm not sure it matters that much either way.